src/Authenticator/EnvironmentAuthenticator.php

```diff
@@ -0,0 +1,127 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         2.10.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Authenticator;
+
+use Authentication\UrlChecker\UrlCheckerTrait;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Form Authenticator
+ *
+ * Authenticates an identity based on the POST data of the request.
+ */
+class EnvironmentAuthenticator extends AbstractAuthenticator
+{
+    use UrlCheckerTrait;
+
+    /**
+     * Default config for this object.
+     * - `fields` The fields to use to identify a user by.
+     * - `loginUrl` Login URL or an array of URLs.
+     * - `urlChecker` Url checker config.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'loginUrl' => null,
+        'urlChecker' => 'Authentication.Default',
+        'attributes' => [],
+    ];
+
+    /**
+     * Checks the fields to ensure they are supplied.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request that contains login information.
+     * @return array|null server params defined by fields.
+     */
+    protected function _getData(ServerRequestInterface $request): ?array
+    {
+        $attributes = $this->_config['attributes'];
+        /** @var array $params */
```

Comment:
Is it annotation really necessary? Return value of `getServerParams()` is always an array.